### PR TITLE
[Metricbeat] GCP cloudsql metadata

### PR DESCRIPTION
**Pull request by ghost**
## What does this PR do?

While working on the GCP CloudSQL integration https://github.com/elastic/integrations/pull/4126 I discovered that there is no way to know which database type (PostgreSQL, MySQL or SQLServer) generated an event.

Adds `cloudsql` metadata.

An event would look like this with the added `cloudsql` labels.

```
 "gcp": {
    "labels": {
      "resource": {
        "region": "us-central",
        "database_id": "elastic-observability:postgres"
      },
      "cloudsql": {
        "version": "14",
        "name": "postgres"
      },
      "metrics": {
        "state": "SUSPENDED"
      }
    },
    "metrics": {
      "database": {
        "instance_state": {
          "value": false
        }
      }
    }
  },
```

@kaiyan-sheng suggested adding the `cloudsql` metadata under `gcp` not under `labels` but the stackdriver metadata only allows access to `ECS` and `Labels`.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Comment by endorama** [comment]
@tommyers-elastic May you have a look at this comment:
> @kaiyan-sheng suggested adding the cloudsql metadata under gcp not under labels but the stackdriver metadata only allows access to ECS and Labels.

and provide your opinion here? We can either try to fix this here, leave it like this for consistency or try to refactor (we would need an exploration phase to understand how this would play out)

**Comment by kaiyan-sheng** [comment]
Sorry I missed the comment here. I'm fine with leaving the fields under `labels` for consistency. I mentioned `gcp` because that's what the metadata are stored for AWS as an example, not a suggestion. But I'd rather have the consistency and no breaking change here.

**Comment by endorama** [comment]
@gpop63 I was reviewing this contribution, there is only one thing that we need to double check: this code uses the same caching logic as `instance` metadata. [We discovered a bug](https://github.com/elastic/beats/issues/32734) in the caching logic there, and we decided to remove the cache as it wasn't been used properly.

Can you please remove the caching logic from this module? It will be reintroduced if the need arise in future updates.

**Comment by gpop63** [comment]
I have a question regarding `cloud.id` and `cloud.name` fields. CloudSQL seems to not have unique ints as instance ids, the id of an instance is a concatenation of the project name and the instance name `{project}:{instance_name}`. Is this ok?
  
```
"cloud": {
  "instance": {
    "id": "elastic-observability:postgres",
    "name": "postgres"
  },
```

**Comment by endorama** [comment]
@gpop63 If that is the GCP provided ID I would move forward with it. We need to make sure the mapping in the data stream is correct though, or there could be errors during ingestion.

File: x-pack/metricbeat/module/gcp/metrics/cloudsql/metadata.go [hunk 1]
```diff
@@ -57,13 +50,13 @@ type cloudsqlMetadata struct {
...
-	regions       []string
-	opt           []option.ClientOption
-	instanceCache *common.Cache
-	logger        *logp.Logger
+	projectID string
+	zone      string
+	region    string
+	regions   []string
+	opt       []option.ClientOption
+	instances map[string]*sqladmin.DatabaseInstance
```
Comment (endorama) [inline x-pack/metricbeat/module/gcp/metrics/cloudsql/metadata.go:59]:
```suggestion
	// NOTE: instances holds data used for all metrics collected in a given period
	// this avoids calling the remote endpoint for each metric, which would take a long time overall
	instances map[string]*sqladmin.DatabaseInstance
```

**Review by endorama — COMMENTED** [review]
Adding a small comment to clarify the role of `instances`.

**Review by kaiyan-sheng — APPROVED** [review]
looks good to me, could you add a changelog entry please? Thank you!

File: x-pack/metricbeat/module/gcp/metrics/cloudsql/metadata.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		opt:       opt,
+		instances: make(map[string]*sqladmin.DatabaseInstance),
+		logger:    logp.NewLogger("metrics-cloudsql"),
+	}, nil
+}
+
+// cloudsqlMetadata is an object to store data in between the extraction and the writing in the destination (to uncouple
+// reading and writing in the same method)
+type cloudsqlMetadata struct {
+	// projectID   string
```
Comment (kaiyan-sheng) [inline x-pack/metricbeat/module/gcp/metrics/cloudsql/metadata.go:38]: should we remove the projectID,  machineType and ts here?

**Comment by kaiyan-sheng** [comment]
> I have a question regarding `cloud.id` and `cloud.name` fields. CloudSQL seems to not have unique ints as instance ids, the id of an instance is a concatenation of the project name and the instance name `{project}:{instance_name}`. Is this ok?
> 
> ```
> "cloud": {
>   "instance": {
>     "id": "elastic-observability:postgres",
>     "name": "postgres"
>   },
> ```

Actually looking at this comment, I don't think we should use `cloud.instance.id` and `cloud.instance.name` for these fields. In ECS, cloud.instance.id is defined as `Instance ID of the host machine.`. I would probably just use `gcp.cloudsql.id` and `gcp.cloudsqp.name` instead here.

**Pull request merged**